### PR TITLE
Add Storage DataProtection integration package

### PR DESCRIPTION
**Pull request by pakrym**
This migrates existing dataprotection key storage extension to track 2 (https://www.nuget.org/packages/Microsoft.AspNetCore.DataProtection.AzureStorage/)

### CODE REVIEW

**Comment by tg-msft** [comment]
Should this be `Azure.Storage.Blobs.AspNetCore.DataProtection` or maybe something more like `Azure.AspNetCore.DataProtection.Blobs`?  +@KrzysztofCwalina

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj]: Can we include a link to the ASP.NET documentation on key stores?  Maybe https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/implementation/key-storage-providers would be best here.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj]: Throw in `key store` as well?

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj]: Have we received an official stamp of approval for this dependency?

File: eng/.docsettings.yml
Comment (tg-msft) [inline eng/.docsettings.yml]: I think most of these bugs reference the `azure-sdk-tools` repo so it would be good to make this something more like `azure-sdk-for-net #9960`.

**Comment by pakrym** [comment]
Naming is something I'm very open to changing. I'm trying to get the package building and ship a preview out of our repo so partners can start using it. APIs and and names are very open for discussion.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs]: I feel like this is the third copy of this code.  Might be worth filing a bug to create a shared source version of it.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs]: Any reason to not always so this?

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs]: Style nit - move this up top so folks can see the definition before it gets used everywhere.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureStorageBlobDataProtectionBuilderExtensions.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureStorageBlobDataProtectionBuilderExtensions.cs]: I guess there's no meaningful case where this would ever be a public access blob, right?  That's the only other case where the Sas would be null with a shared key.

Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureStorageBlobDataProtectionBuilderExtensions.cs]: Would it actually be nicer to throw if there's no Sas and make users supply the `DefaultAzureCredential` themselves in the other overload that takes a token credential?  I don't have a good feel for how much magic is the right amount of magic in the ASP.NET world.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureStorageBlobDataProtectionBuilderExtensions.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureStorageBlobDataProtectionBuilderExtensions.cs]: I find it a little confusing that we've got helpers for Sas/TokenCredential but not SharedKeyCredential/ConnectionString.  I get that users could instantiate it themselves and pass a BlobClient in to the other overload, but I'm not sure it will be as obvious to others.  I'm guessing this is a design decision to guide people toward better auth choices?  I think it's fine to Preview but would be worth looking to see if people struggle here.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/tests/MockResponse.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/tests/MockResponse.cs]: Any reason we're not pulling it in as shared source?

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/tests/AzureBlobXmlRepositoryTests.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/tests/AzureBlobXmlRepositoryTests.cs]: Can we add one to test what happens when a blob does exist there but it's not ASP.NET keys?

**Review by tg-msft — APPROVED** [review]

**Comment by tg-msft** [comment]
I feel like we should structure the naming as `Azure.AspNetCore.*` because this is more relevant to ASP.NET users than other Storage users.

**Comment by pakrym** [comment]
> I feel like we should structure the naming as Azure.AspNetCore.* because this is more relevant to ASP.NET users than other Storage users.

There are other packages that need to have Extension in the name and I'm worried that `Azure.Extensions.*` might be a little to generic.

I like that the names are way shorted the way you propose.

File: eng/.docsettings.yml
Comment (pakrym) [reply eng/.docsettings.yml]: No, they are all in azure-sdk-for-net

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/tests/MockResponse.cs
Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/tests/MockResponse.cs]: Leftovers from copying it from AspNetCore repo.

File: eng/.docsettings.yml
Comment (tg-msft) [reply eng/.docsettings.yml]: Sorry, I had it [backwards](https://github.com/Azure/azure-sdk-for-net/blob/e97701b729ebf54c4095c1af57c887695b32a0fc/eng/.docsettings.yml#L133).  You're correct.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs
Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs]: I know we have exponential backoff in couple places but haven't seen random back off anywhere else. Do you have an example?

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs
Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs]: I find it a little bit easier when private types are at the end of the file.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj
Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj]: Haven't asked for it, seems like if we are shipping a plugin for something we got to be able to reference that something.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/AzureBlobXmlRepository.cs]: I was thinking of https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Core/BasicRetryPolicy.cs#L115, but that is sufficiently more complicated.

File: sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj]: We've had this discussion with other partners (i.e, I'm shipping an Avro feature so I've got to be able to reference the Avro parser) and it'd be good if our team [followed the same rules](https://azure.github.io/azure-sdk/general_implementation.html#general-dependencies-approved-only).  I can't imagine you wouldn't get it approved though.

Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj]: Sounds good.

Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj]: Would you prefer getting the approval before merging the PR?

Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs.AspNetCore.DataProtection/src/Azure.Storage.Blobs.AspNetCore.DataProtection.csproj]: No, so long as @KrzysztofCwalina gives a thumbs up before we ship any Previews.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/Azure.Security.KeyVault.Secrets.Extensions.Configuration.csproj
Comment (heaths) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/Azure.Security.KeyVault.Secrets.Extensions.Configuration.csproj]: Why on the test projects?

File: eng/Packages.Data.props
Comment (heaths) [inline eng/Packages.Data.props]: Doesn't this increase the risk of dependencies getting out of date? I don't see any benefit here since the compiler will elide any references that aren't needed and, in fact, these don't actually add package references anyway - just update existing ones with a version.

**Review by heaths — APPROVED** [review]
Signing off with comments. Only reviewed general and KV files.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/Azure.Security.KeyVault.Secrets.Extensions.Configuration.csproj
Comment (pakrym) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets.Extensions.Configuration/src/Azure.Security.KeyVault.Secrets.Extensions.Configuration.csproj]: Good point, I'll move some packages under IsTestProject condition instead.

File: eng/Packages.Data.props
Comment (pakrym) [reply eng/Packages.Data.props]: This condition is to prevent "normal" azure SDKs from taking dependency on "extensions" packages that are intended only for a few selected plugin packages.

**Comment by pakrym** [comment]
Filed https://github.com/Azure/azure-sdk-for-net/issues/10003 to follow up on extension methods.

File: eng/Packages.Data.props
Comment (weshaggard) [inline eng/Packages.Data.props]: This is already in the larger package list and set to 12.2.0.

File: eng/Packages.Data.props
Comment (weshaggard) [inline eng/Packages.Data.props]: All of the extension are already in the general item group above as well. While this override some it doesn't eliminate others. Part my issue with this change is trying to avoid duplicate versions across our different package sets. Perhaps we should figure out a way to ensure we don't have duplicate items across the different package types where there are different conflicting versions. Ideally we shouldn't find a package listed more then once in this file.

File: eng/Packages.Data.props
Comment (pakrym) [reply eng/Packages.Data.props]: Not sure what's up with this. I just moved this reference around.

File: eng/Packages.Data.props
Comment (pakrym) [reply eng/Packages.Data.props]: Let me see who references those and try to clean them up.

Comment (weshaggard) [reply eng/Packages.Data.props]: They might be referenced by the track 1 libraries. At a min we should probably make the lists mutually exclusive (i.e. make the larger top list `'$(IsClientLibrary)' != 'true').

Comment (pakrym) [reply eng/Packages.Data.props]: Making them exclusive might involve a lot of churn that I don't want to include into this PR. I moved all Extensions references to test project only group.

Comment (pakrym) [reply eng/Packages.Data.props]: This file is also used as the baseline for ApiCompat.

Comment (weshaggard) [reply eng/Packages.Data.props]: We might need to start using a separate file for ApiCompat.

**Comment by KrzysztofCwalina** [comment]
> Should this be Azure.Storage.Blobs.AspNetCore.DataProtection or maybe something more like Azure.AspNetCore.DataProtection.Blobs?

I think Azure.AspNet.DataProtection.Blobs is better. The feature is only interesting to people who use ASP.NET, and so it should be grouped in an "asp.net" group. I don't think we should sprinkle these types all over Azure SDK namespaces.

**Comment by pakrym** [comment]
> I think Azure.AspNet.DataProtection.Blobs is better. The feature is only interesting to people who use ASP.NET, and so it should be grouped in an "asp.net" group. I don't think we should sprinkle these types all over Azure SDK namespaces.

Sounds good, but I'd like to change one this and use AspNetCore instead of AspNet.

**Pull request merged**
